components/function-controller/design/extend_function_spec_proposal.md

```diff
@@ -0,0 +1,392 @@
+# Extending CRD spec for Serverless v1alpha2
+
+## Summary
+
+The current Serverless API allows for limited configuration of the generated Function's deployment. 
+Currently users can only use ENVs to pass 3rd party service credentials but it doesn't allow for volume mounted secrets (which becomes the industry standard for [service bindings](https://servicebinding.io/application-developer/)).
+Moreover users have no control over the annotations that are applied on the function runtime pod. This excludes functions pods from features enabled by annotations (i.e custom log parsers via `fluentbit.io/parser: my-regex-parser`).
+
+
+
+## Motivation
+
+Give Serverless users the ability to:
+- Configure volume mounted secrets (and config maps) for Function's subresources.
+- Configure labels and annotations for the  Function runtime pod.
+
+### Goals
+
+- Add more flexibility to the Serverless API - enable volume mounted secrets and pod annotations.
+- Organise spec attributes belonging to runtime and build-time configuration (?)
+- Provide a rollout plan to implement this functionality in small increments and avoid the need to roll out a new API version
+
+### Non-Goals
+- Define a new API version specification
+
+## Discussion points
+
+### Runtime, build-time separation
+
+Since Function CR is managing two workloads ( Deployment for runtime and Job for build-time ) we need to separate them in the spec in order to make it clear where do the mounts and annotations belong. 
+
+We can either make a clear separation, i.e:
+
+```yaml 
+spec:
+  source:
+  runtimeSpec:
+    # mounts
+    # resources
+    # envs
+    # metadata
+  buildSpec:
+    # resources
+    # envs (?)
+    # metadata (?)
+```
+
+Alternatively we could promote runtime fields to the root (as the belong to kind: Function) and extract only the build-time fields 
+
+```yaml 
+spec:
+  source:
+  # mounts
+  # resources
+  # envs
+  # metadata
+  buildSpec:
+    # resources
+    # envs (?)
+    # metadata (?)
+```
+
+### Mounts - own structure or k8s inherited
```

Comment:
I prefer providing the user with more flexibility and exposing the full volume API, but if we can't, the serviceBinding option sound good.

Reply:
I discussed it with @pbochynski 

we could separate those cases.
We definitely need a convenient way to enable service bindings.
Those use natively  k8s secrets ( i.e created via BTP service binding CRs ) and leverage dedicated SDKs compliant with  service binding recommendation

volume mounts is more generic but would be less convinient for devs focused on the main use case ( service bindings )

Volume mounts could be a separate feature in our spec ( see last "compromise" option )